heka/plugins/sqs3/sqs3_input.go

```diff
@@ -0,0 +1,178 @@
+        AwsRegion: "us-east-1",
+    }
+}
+
+func (input *Sqs3Input) Init(config interface{}) error {
+    conf := config.(*Sqs3InputConfig)
+    input.Sqs3InputConfig = conf
+
+    input.sqs = sqs.New(session.New(), aws.NewConfig().WithRegion(input.AwsRegion))
+    input.s3 = s3.New(session.New(), aws.NewConfig().WithRegion(input.AwsRegion))
+
+    queue_url, err := get_queue(input.sqs, input.SqsQueue)
+    if err != nil { return err }
+    input.queue_url = queue_url
+
+    input.stop = make(chan bool)
+    return nil
+}
+
+func (input *Sqs3Input) Stop() {
+    close(input.stop)
+}
+
+func (input *Sqs3Input) Run(runner pipeline.InputRunner, helper pipeline.PluginHelper) error {
+    splitterRunner := runner.NewSplitterRunner("")
+    defer splitterRunner.Done()
+
+    for {
+        select {
+        case <-input.stop:
+            return nil
+        default:
+        }
+
+        receipt_handle, bucket, key, err := receive_from_queue(input.sqs, input.queue_url)
+        if err != nil {
+            runner.LogError(fmt.Errorf("Error reading queue: %s", err.Error()))
+            continue
+        }
+
+        o, err := get_object(input.s3, bucket, key)
+        if err != nil {
+            runner.LogError(fmt.Errorf("Error opening s3object: %s", err.Error()))
+            if aws_err, ok := err.(awserr.Error); ok {
+                if aws_err.Code() == "NoSuchBucket" || aws_err.Code() == "NoSuchKey" {
+                    delete_message(input.sqs, input.queue_url, receipt_handle)
+                }
+            }
+            continue
+        }
+        runner.LogMessage(fmt.Sprintf("Sqs3Input: found s3://%s/%s", *bucket, *key))
+        for err == nil {
+            err = splitterRunner.SplitStream(o, nil)
+            if err == io.EOF {
+                delete_message(input.sqs, input.queue_url, receipt_handle)
+                break
+            } else if err != nil {
+                runner.LogError(fmt.Errorf("Error reading file: %s", err.Error()))
+            }
+        }
+        o.Close()
+    }
+}
+
+func init() {
+    pipeline.RegisterPlugin("Sqs3Input", func() interface{} {
+        return new(Sqs3Input)
+    })
+}
+
+// helper functions
+
+type SqsBody struct {
+    Records []struct {
+        EventName string
+        S3        struct {
+            Bucket struct {
+                Name string
+            }
+            Object struct {
+                Key  string
+                Size int
+            }
+        }
+    }
+}
+
+func get_queue(svc *sqs.SQS, queue string) (*string, error) {
+    params := &sqs.GetQueueUrlInput{
+        QueueName: aws.String(queue),
+    }
+    resp, err := svc.GetQueueUrl(params)
+    if err != nil { return nil, err }
+    return resp.QueueUrl, nil
+}
+
+func receive_from_queue(svc *sqs.SQS, queue_url *string) (*string, *string, *string, error) {
+    // get sqs message
+    params := &sqs.ReceiveMessageInput{
+        QueueUrl: queue_url,
+        MaxNumberOfMessages: aws.Int64(1),
```

Comment:
It might be good to add a `WaitTimeSeconds` setting here to put an upper bound on how long you wait for messages. I believe that it will wait for a message forever by default, which would likely cause trouble when trying to shut down the plugin if the queue is empty.